packages/analytics/matomo/usePaq.ts

```diff
@@ -1,6 +1,14 @@
+declare global {
+   interface Window {
+      _paq: any;
+   }
+}
+
 export function usePaq(): any[] | undefined {
    if (typeof window === 'undefined') {
       return undefined;
    }
-   return (window as any)._paq;
+
+   var _paq = (window._paq = window._paq || []);
```

Comment:
Although this works, it seems to spy a larger bug to me.

We are trying to use matomo *before* we have even included matomo on the page.

Reply:
I think it's okay. The Matomo tracking object takes whatever you've added the to the `_paq` array and adopts it and the tracking object has a `push` method that sort of masks over it.

Reply:
That's always been the way that analytics platforms work (GA, matomo, pwik, ...). They use an array interface so `window._paq` starts as an array and matomo then (when it loads) replaces that with something that exposes a `push()` method just like array does and at the same time consumes the already pushed calls.

Reply:
Agreed, but I think this indicates matomo needs to be included higher on in the layout template IMO

Reply:
Next.js [recommends](https://nextjs.org/docs/api-reference/next/script#afterinteractive) that analytics scripts be loaded after the page becomes interactive. We're currently following this suggestion, but we could experiment with running the script `beforeInteractive` if it's important to us.

https://github.com/iFixit/react-commerce/blob/34257205922f58e7628a0f9a559e7dfd6d1febd2/frontend/components/analytics/Matomo.tsx#L19

Reply:
Oh man, I wasn't aware of the nuance of the Next.js script loading here.

If I (briefly) ignore how next.sj works. I totally agree the rendered script should be something like this: `<script async="" src="...matomo.js"></script>`

This script should be de-prioritized for page loading, but high enough in the dom that the parsing (and background downloading) happens as soon as possible.

My apologies here, but I am not well versed in how the script component exactly ends up rendered in this case.

For clarity here:

It is totally fine for us to push the array before we have included the matomo script tag on the page. The fact that we hit a case where the global array was not yet defined seems to indicate that we have code that ran before the matomo script in the dom.

This pull will address the problem of the array being undefined, which is great. My claim is we probably can move the matomo script rendering higher up in the dom tree.